cmd/sign-executable/sign-executable.go

```diff
@@ -0,0 +1,291 @@
+package main
+
+import (
+	"archive/zip"
+	"bytes"
+	"context"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"path"
+
+	"github.com/pkg/errors"
+	"github.com/urfave/cli"
+)
+
+const (
+	getClientSubcommand  = "get-client"
+	signSubcommand       = "sign"
+	notaryClientFilename = "macnotary"
+)
+
+func zipFile(inputFilePath, destinationPath string) error {
+	zipFile, err := os.Create(destinationPath)
+	if err != nil {
+		return errors.Wrap(err, "creating zip file")
+	}
+	defer zipFile.Close()
+
+	zipWriter := zip.NewWriter(zipFile)
+	defer zipWriter.Close()
+
+	inputFile, err := os.Open(inputFilePath)
+	if err != nil {
+		return errors.Wrapf(err, "opening input file '%s'", inputFilePath)
+	}
+	defer inputFile.Close()
+
+	fileInfo, err := inputFile.Stat()
+	if err != nil {
+		return errors.Wrap(err, "getting info for input file")
+	}
+
+	header, err := zip.FileInfoHeader(fileInfo)
+	if err != nil {
+		return errors.Wrap(err, "making header from file info")
+	}
+	header.Method = zip.Deflate
+
+	writer, err := zipWriter.CreateHeader(header)
+	if err != nil {
+		return errors.Wrap(err, "writing header to zip file")
+	}
+	_, err = io.Copy(writer, inputFile)
+	return errors.Wrap(err, "copying input file into zip file")
+}
+
+func zipFileToFile(file *zip.File, destination string) error {
+	fileReader, err := file.Open()
+	if err != nil {
+		return errors.Wrap(err, "opening zip file")
+	}
+	defer fileReader.Close()
+
+	contents, err := io.ReadAll(fileReader)
+	if err != nil {
+		return errors.Wrap(err, "reading zip file")
+	}
+
+	return errors.Wrap(os.WriteFile(destination, contents, file.Mode()), "writing zip contents to file")
+}
+
+func extractFileFromZip(zipBytes []byte, targetFile, destination string) error {
+	reader := bytes.NewReader(zipBytes)
+	zipReader, err := zip.NewReader(reader, reader.Size())
+	if err != nil {
+		return errors.Wrap(err, "getting zip reader")
+	}
+
+	targetFound := false
+	for _, file := range zipReader.File {
+		if path.Base(file.Name) != targetFile {
+			continue
+		}
+		if file.FileInfo().IsDir() {
+			continue
+		}
+		targetFound = true
+		if err = zipFileToFile(file, destination); err != nil {
+			return errors.Wrap(err, "extracting zip file")
+		}
+	}
+
+	if !targetFound {
+		return errors.Errorf("target file not in the archive")
+	}
+
+	return nil
+}
+
+func downloadZip(ctx context.Context, downloadURL string) ([]byte, error) {
```

Comment:
I did take archiver for a whirl, but I ended up back where I started.
I mean, I could have used it, but it would have been an effort to bend the input/results to what I was looking for.

Reply:
Specifically, the format of the client release is with two members in the archive:
* linux_amd64/
* linux_amd64/macnotary
To use [Extract](https://pkg.go.dev/github.com/mholt/archiver/v3#Zip.Extract) I'd need to pass in the second path and then move the actual binary out of the directory to where I want it. It's not insurmountable, but annoying enough that I'm inclined to use the zip package directly. Reading like [this](https://pkg.go.dev/github.com/mholt/archiver/v3#example-Zip-StreamingRead) would work but seems roughly the same amount of custom logic as using the zip package.